src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState.cs

```diff
@@ -390,5 +401,17 @@ public virtual async Task<VersionStamp> GetTopLevelChangeTextVersionAsync(Cancel
             TextAndVersion textAndVersion = await this.textAndVersionSource.GetValueAsync(cancellationToken).ConfigureAwait(false);
             return textAndVersion.Version;
         }
+
+        internal sealed class DefaultDocumentServiceProvider : IDocumentServiceProvider
```

Comment:
does this need to be internal?

Reply:
if you want to see full design, please see the original PR. sam is breaking that PRs to multiple ones. each small PRs to deal with 1 concept a time.

it only works for people who knows how all connects at the end.

having PR on ILanguageServiceFactory/ILanguageService and asking how HostLanguageService/MefLanguageService and HostWorkspaceServices/MefWorkspaceServices/IWorkspaceService and Workspace are all interacting each others at the end are too broad question.

Reply:
> if you want to see full design, please see the original PR. sam is breaking that PRs to multiple ones. each small PRs to deal with 1 concept a time.

I did see that PR.  That was the PR where i asked for the individual pieces to be pulled out :D 

> there seemed to eb a lot of different pieces of work wrapped into a single PR. It would be much nicer if we could just extract and PR the individual pieces.

:) :)

Even looking at the full PR, there were a lot of questions and concerns i had.  The benefit of breaking htings out is that we can see the pieces independently, and see if they maje sense, esp. given the types of scenarios you are trying to solve in #30733

> are all interacting each others at the end are too broad question.

I disagree.  This is a very specific question for this PR.  If this doesn't need to be internal here, then make it private.  Then, in the PRs that follow, you can make it internal.  It will then be clear *from those PRs* why this would need to be internal and not private.